### PR TITLE
refactor(source): enhance parquet file source

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
This PR primarily focuses on the conversion of the `Timestamp` type. The Arrow `Timestamp` includes four types:
```

pub enum TimeUnit {
    /// Time in seconds.
    Second,
    /// Time in milliseconds.
    Millisecond,
    /// Time in microseconds.
    Microsecond,
    /// Time in nanoseconds.
    Nanosecond,
}
```
but previously we only handled one `Microsecond`. This pr makes the comparison of data types more reasonable:
- when doing data type check at column pruning, make `risingwave_common::types::DataType::Timestamp` match `arrow_schema_iceberg::DataType::Timestamp(_, _)`.
- when convert arrow field to rw field, add `Nanosecond`, `Millisecond`, `Second` type.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 1]
```diff
@@ -285,3 +285,85 @@ pub fn extract_valid_column_indices(
...
+            if let arrow_schema_iceberg::DataType::List(field_ref) = arrow_data_type {
+                let inner_rw_type = inner_type.clone();
+                let inner_arrow_type = field_ref.data_type();
+                is_data_type_matching(&inner_rw_type, inner_arrow_type)
+            } else {
+                false
+            }
+        }
+        risingwave_common::types::DataType::Map(_) => {
+            // Directly return false for Map types
```
Comment (wcy-fdu) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:364]: I think `Map` type in RW and arrow is quite different, the former is `pub struct MapType(Box<(DataType, DataType)>);` while the latter is `Arc<Field>`, so I just reject this type comparison. Pls correct me if I was wrong.

Comment (chenzl25) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:364]: cc @xxchan

Comment (xxchan) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:364]: No, they have the same representation. Arrow's `Arc<Field>` is just a struct with 2 fields. 

Please check `src/common/src/array/arrow/arrow_impl.rs` for the conversion code.

File: src/common/src/array/arrow/arrow_impl.rs [hunk 2]
```diff
@@ -692,6 +709,33 @@ pub trait FromArrow {
...
+    ) -> Result<ArrayImpl, ArrayError> {
+        Ok(ArrayImpl::Timestamp(array.into()))
+    }
+
+    fn from_timestampms_some_array(
+        &self,
+        array: &arrow_array::TimestampMillisecondArray,
+    ) -> Result<ArrayImpl, ArrayError> {
+        Ok(ArrayImpl::Timestamptz(array.into()))
+    }
```
Comment (chenzl25) [inline src/common/src/array/arrow/arrow_impl.rs:737]: I am curious, we don't need to do any conversion between these types? All of them can be implemented in the same way?

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 3]
```diff
@@ -285,3 +285,85 @@ pub fn extract_valid_column_indices(
...
+///
+/// This function compares the `DataType` from RisingWave with the `DataType` from
+/// Parquet file, returning `true` if they are compatible. Specifically, for `Timestamp`
+/// types, it ensures that any of the four `TimeUnit` variants from Parquet
+/// (i.e., `Second`, `Millisecond`, `Microsecond`, and `Nanosecond`) can be matched
+/// with the corresponding `Timestamp` type in RisingWave.
+pub fn is_data_type_matching(
+    rw_data_type: &risingwave_common::types::DataType,
+    arrow_data_type: &arrow_schema_iceberg::DataType,
+) -> bool {
```
Comment (xxchan) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:299]: Why do we need to add a new function for the check? This leads to duplicated code, and might lead to inconsistency.

Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:299]: Previously we do column pruning(aka judge the rw schema matches parquet file schame) by converting rw column to arrow `Field`, and the compare two arrow `Field`. However, the problem here is that the timestamp type of rw will only become the `Timestamp(Microsecond)` of arrow during the conversion process, this cause a mismatch between the timestamp of rw and the `Timestamp(Nanosecond)`, `Timestamp(Millisecond)`, etc. of arrow.

In short, what we need is a relatively loose type comparison.

Comment (xxchan) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:299]: Do you mean 1 rw type can correspond to N arrow types?

If so, can we test type matching by converting arrow type to rw type?

Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:299]: Yes, I think we can use from_field, but need to add the missing type first.
https://github.com/risingwavelabs/risingwave/blob/f559d3468a16f0a15e429dfb55276b1a07c814e3/src/common/src/array/arrow/arrow_impl.rs#L502-L514

File: src/common/src/array/arrow/arrow_impl.rs (hunk 2 above)
Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:744]: No, just added such logic and verify it works. I constructed a parquet file containing timestamps of four time units, then imported it into RisingWave and it was parsed successfully:
image

File: src/common/src/array/arrow/arrow_impl.rs [hunk 4]
```diff
@@ -841,7 +907,43 @@ macro_rules! converts {
...
         }
     };
+
+    ($ArrayType:ty, $ArrowType:ty, $time_unit:expr, @map) => {
+
+        impl From<&$ArrayType> for $ArrowType {
+            fn from(array: &$ArrayType) -> Self {
+                array.iter().map(|o| o.map(|v| v.into_arrow())).collect()
+            }
+        }
```
Comment (chenzl25) [inline src/common/src/array/arrow/arrow_impl.rs:917]: It seems we need to have a method `into_arrow_with_unit()` as well.

File: src/common/src/array/arrow/arrow_impl.rs [hunk 5]
```diff
@@ -854,21 +956,35 @@ converts!(Utf8Array, arrow_array::StringArray);
...
 trait FromIntoArrow {
     /// The corresponding element type in the Arrow array.
     type ArrowType;
+    /// The timestamp type used to distinguish different time units, only utilized when the Arrow type is a timestamp.
+    type TimestampType;
     fn from_arrow(value: Self::ArrowType) -> Self;
     fn into_arrow(self) -> Self::ArrowType;
+    /// Used for converting timestamp types and will not be used in conversions of other types.
+    fn from_arrow_with_unit(value: Self::ArrowType, time_unit: Self::TimestampType) -> Self;
 }
```
Comment (xxchan) [inline src/common/src/array/arrow/arrow_impl.rs:983]: Generally speaking, for type conversion, schema (i.e., type information) should be provided, since data itself may not be self-contained.

```rust
fn convert(v_from, t_from, t_to) -> v_to 
```
`t_from` & `t_to` are not always needed though

e.g., our Avro conversion

https://github.com/risingwavelabs/risingwave/blob/9a32e75f227f837a6f16fb3afe159c1f8af2ad5b/src/connector/codec/src/decoder/avro/mod.rs#L89-L93

Comment (xxchan) [reply src/common/src/array/arrow/arrow_impl.rs:983]: Therefore, `_with_unit` looks a little strange to me. Haven't checked the details of Arrow yet.

Comment (xxchan) [reply src/common/src/array/arrow/arrow_impl.rs:983]: cc @xiangjinwu @BugenZhao might also be able to comment

File: src/common/src/array/arrow/arrow_impl.rs (hunk 4 above)
Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:917]: Here, the main consideration is compatibility. Previously, user sink timestamps with only a microsecond precision. If we change this to allow sinking in seconds, milliseconds, etc., downstream users might encounter multiple columns added. In the current implementation, zeros are only added when precision is insufficient, and there are no correctness issues. I suggest we maintain the current situation.

File: src/common/src/array/arrow/arrow_impl.rs (hunk 5 above)
Comment (xiangjinwu) [reply src/common/src/array/arrow/arrow_impl.rs:983]: https://docs.rs/arrow-schema/53.2.0/arrow_schema/enum.DataType.html
If we need `_with_unit` to pass extra info from `DataType`, how about the following?
* `IntervalUnit`
* `i32` length for `FixedSizeBinary`
* precision and scale for `Decimal{128,256}`

To repeat, `DataType` should be provided as data itself may not be self-contained.

File: e2e_test/s3/fs_sink.py [hunk 6]
```diff
@@ -147,7 +166,7 @@ def _table():
     print('Sink into s3 in parquet encode...')
     # Execute a SELECT statement
     cur.execute(f'''CREATE TABLE test_parquet_sink_table(
-        id bigint primary key,
+        id bigint primary key,\
```
Comment (stdrc) [inline e2e_test/s3/fs_sink.py:169]: Does this make any difference?

File: src/common/src/array/arrow/arrow_impl.rs [hunk 7]
```diff
@@ -841,7 +907,43 @@ macro_rules! converts {
...
+                    .map(|o| {
+                        o.map(|v| {
+                            <<$ArrayType as Array>::RefItem<'_> as FromIntoArrow>::from_arrow(v)
+                        })
+                    })
+                    .collect()
+            }
+        }
+
+    };
```
Comment (stdrc) [inline src/common/src/array/arrow/arrow_impl.rs:944]: This looks very specialized. What about put it in a new macro, like `converts_timestamp`?

File: src/common/src/array/arrow/arrow_impl.rs [hunk 8]
```diff
@@ -854,21 +956,37 @@ converts!(Utf8Array, arrow_array::StringArray);
...
+
 converts!(IntervalArray, arrow_array::IntervalMonthDayNanoArray, @map);
 converts!(SerialArray, arrow_array::Int64Array, @map);
 
 /// Converts RisingWave value from and into Arrow value.
 trait FromIntoArrow {
     /// The corresponding element type in the Arrow array.
     type ArrowType;
+    /// The timestamp type used to distinguish different time units, only utilized when the Arrow type is a timestamp.
+    type TimestampType;
```
Comment (stdrc) [inline src/common/src/array/arrow/arrow_impl.rs:978]: Why is this needed? I guess it's too intrusive to put `timestamp` related things in every type?

Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:978]: We need to distinguish the four timestamp types of arrow, instead of just using one i64 to represent them.

File: src/common/src/array/arrow/arrow_impl.rs [hunk 9]
```diff
@@ -512,6 +513,12 @@ pub trait FromArrow {
             Time64(Microsecond) => DataType::Time,
             Timestamp(Microsecond, None) => DataType::Timestamp,
             Timestamp(Microsecond, Some(_)) => DataType::Timestamptz,
+            Timestamp(Second, None) => DataType::Timestamp,
+            Timestamp(Second, Some(_)) => DataType::Timestamptz,
+            Timestamp(Millisecond, None) => DataType::Timestamp,
+            Timestamp(Millisecond, Some(_)) => DataType::Timestamptz,
+            Timestamp(Nanosecond, None) => DataType::Timestamp,
+            Timestamp(Nanosecond, Some(_)) => DataType::Timestamptz,
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:521]:
```
Timestamp(_, None) => DataType::Timestamp,
Timestamp(_, Some(_)) => DataType::Timestamptz,
```

File: src/common/src/array/arrow/arrow_impl.rs (hunk 2 above)
Comment (hzxa21) [reply src/common/src/array/arrow/arrow_impl.rs:744]: I also don't get why we need to add different methods to convert timestamp in different units. IIUC, the current conversion [logic ](https://github.com/risingwavelabs/risingwave/blob/fdeeb2e6c358fe8f04d1fa29ef5384e75ae2e3f8/src/common/src/array/arrow/arrow_impl.rs#L944)won't lose any precision. What happened if we just add different units of the timestamp type to `converts` macro and rely on the original [implementation ](https://github.com/risingwavelabs/risingwave/blob/fdeeb2e6c358fe8f04d1fa29ef5384e75ae2e3f8/src/common/src/array/arrow/arrow_impl.rs#L942) to do the conversion?
```
converts!(TimestampArray, arrow_array::TimestampMillisecondArray, @map);
converts!(TimestampArray, arrow_array::TimestampSecondArray, @map);
converts!(TimestampArray, arrow_array::TimestampNanosecondArray, @map);
```

Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:744]: In our previous code, we only handled timestamps in microseconds, so this conversion is only correct for `timestamp(MicroSecond, _)`. As you mentioned in the [logic](https://github.com/risingwavelabs/risingwave/blob/fdeeb2e6c358fe8f04d1fa29ef5384e75ae2e3f8/src/common/src/array/arrow/arrow_impl.rs#L944), the seconds part is `% 1_000_000`, and the nanoseconds part is also `% 1_000_000`. If all four time units are converted using this logic, it will yield incorrect results, as follows:
image

Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:744]: The correct approach should be as shown in the PR, where different handling is done based on the time unit. For example, for milliseconds, it should be `%1000` instead of `%100000`:
```
 TimeUnit::Millisecond => Timestamp(
                DateTime::from_timestamp((value / 1_000) as _, (value % 1_000 * 1000000) as _)
                    .unwrap()
                    .naive_utc(),
            ),
```

So I think specifying the time unit is essential.

File: src/common/src/array/arrow/arrow_impl.rs (hunk 9 above)
Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:521]: prefer to keep time unit designated as a reminder, there are four types of timeunit that need to be handled. Previously, other types were ignored because of the use of `_`.

Comment (hzxa21) [reply src/common/src/array/arrow/arrow_impl.rs:521]: Previously we use concrete type `Microsecond` instead of `_`. I don't have a strong opinion on this so listing the types explicitly here is also okay. I am more concerned about having separate logics to handle the different time unit separately given that our original implementation won't lose any precision. See my comment below.

Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:521]: explain in https://github.com/risingwavelabs/risingwave/pull/19221#discussion_r1835933733

**Comment by wcy-fdu** [comment]
In order not to invade the `FromToArrow` trait, the timestamp conversion is split into another trait `FromIntoArrowWithUnit`, and a new macro `converts_with_timeunit!` is also added, which is much clearer than before. As for why it is necessary to formulate a time unit, see https://github.com/risingwavelabs/risingwave/pull/19221#discussion_r1835933733

File: src/common/src/array/arrow/arrow_impl.rs [hunk 10]
```diff
@@ -692,6 +716,33 @@ pub trait FromArrow {
...
+        &self,
+        array: &arrow_array::TimestampSecondArray,
+    ) -> Result<ArrayImpl, ArrayError> {
+        Ok(ArrayImpl::Timestamp(array.into()))
+    }
+    fn from_timestampsecond_some_array(
+        &self,
+        array: &arrow_array::TimestampSecondArray,
+    ) -> Result<ArrayImpl, ArrayError> {
+        Ok(ArrayImpl::Timestamp(array.into()))
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:729]: Timestamptz?

File: src/common/src/array/arrow/arrow_impl.rs [hunk 11]
```diff
@@ -936,34 +1059,80 @@ impl FromIntoArrow for Time {
...
-    fn from_arrow(value: Self::ArrowType) -> Self {
-        Timestamp(
-            DateTime::from_timestamp((value / 1_000_000) as _, (value % 1_000_000 * 1000) as _)
+    fn from_arrow_with_unit(value: Self::ArrowType, time_unit: Self::TimestampType) -> Self {
+        match time_unit {
+            TimeUnit::Second => {
+                Timestamp(DateTime::from_timestamp(value as _, 0).unwrap().naive_utc())
+            }
+            TimeUnit::Millisecond => Timestamp(
+                DateTime::from_timestamp((value / 1_000) as _, (value % 1_000 * 1000000) as _)
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:1072]: I think you can use [DateTime::from_timestamp_millis](https://doc.servo.org/chrono/struct.DateTime.html#method.from_timestamp_millis) here.

File: src/common/src/array/arrow/arrow_impl.rs [hunk 12]
```diff
@@ -936,34 +1059,80 @@ impl FromIntoArrow for Time {
...
+                    .unwrap()
+                    .naive_utc(),
+            ),
+            TimeUnit::Microsecond => Timestamp(
+                DateTime::from_timestamp((value / 1_000_000) as _, (value % 1_000_000 * 1000) as _)
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:1077]: I think you can use `DateTime::from_timestamp_micros` here.

File: src/common/src/array/arrow/arrow_impl.rs [hunk 13]
```diff
@@ -936,34 +1059,80 @@ impl FromIntoArrow for Time {
...
+                    .unwrap()
+                    .naive_utc(),
+            ),
+            TimeUnit::Nanosecond => Timestamp(
+                DateTime::from_timestamp(
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:1082]: I think you can use `DateTime::from_timestamp_nanos` here.

File: src/common/src/array/arrow/arrow_impl.rs [hunk 14]
```diff
@@ -936,34 +1059,80 @@ impl FromIntoArrow for Time {
...
+            TimeUnit::Microsecond => self
+                .0
+                .signed_duration_since(NaiveDateTime::default())
+                .num_microseconds()
+                .unwrap(),
+            TimeUnit::Nanosecond => self
+                .0
+                .signed_duration_since(NaiveDateTime::default())
+                .num_nanoseconds()
+                .unwrap(),
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:1111]: I think you can use `DateTime::timestamp_xxx()` here: https://doc.servo.org/chrono/struct.DateTime.html#method.timestamp_micros

Comment (hzxa21) [reply src/common/src/array/arrow/arrow_impl.rs:1099]: This is not resolved.

File: src/common/src/array/arrow/arrow_impl.rs [hunk 15]
```diff
@@ -584,12 +590,30 @@ pub trait FromArrow {
...
+            Timestamp(Millisecond, None) => {
+                self.from_timestampms_array(array.as_any().downcast_ref().unwrap())
+            }
+            Timestamp(Millisecond, Some(_)) => {
+                self.from_timestampms_some_array(array.as_any().downcast_ref().unwrap())
+            }
             Timestamp(Microsecond, None) => {
                 self.from_timestampus_array(array.as_any().downcast_ref().unwrap())
             }
             Timestamp(Microsecond, Some(_)) => {
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:608]: Not directly related to this PR but I wonder why we ignore the timezone (`Some(_)`) from arrow timestamp when constructing `Timestamptz`

File: src/common/src/array/arrow/arrow_impl.rs [hunk 16]
```diff
@@ -936,34 +1058,69 @@ impl FromIntoArrow for Time {
...
     }
 
-    fn into_arrow(self) -> Self::ArrowType {
-        self.timestamp_micros()
+    fn into_arrow_with_unit(self, time_unit: Self::TimestampType) -> Self::ArrowType {
+        match time_unit {
+            TimeUnit::Second => self.timestamp(),
+            TimeUnit::Millisecond => self.timestamp_millis(),
+            TimeUnit::Microsecond => self.timestamp_micros(),
+            TimeUnit::Nanosecond => self.timestamp_nanos().unwrap_or_default(),
```
Comment (hzxa21) [inline src/common/src/array/arrow/arrow_impl.rs:1122]: Should we use `unwrap` instead of `unwrap_or_default`?

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

File: src/common/src/array/arrow/arrow_impl.rs (hunk 15 above)
Comment (hzxa21) [reply src/common/src/array/arrow/arrow_impl.rs:608]: Introduced in #17201

File: src/common/src/array/arrow/arrow_impl.rs (hunk 16 above)
Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:1122]: I think it's better not to panic when I encounter a problem with the type conversion? It's better to convert to 0 by mistake than to panic a cluster

Comment (hzxa21) [reply src/common/src/array/arrow/arrow_impl.rs:1122]: If our `TimestampTz` is in microsecond unit, I think we should never overflow?

Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:1122]: yes, `microsecond checked_mul(1_000)` is safe, then `unwrap()` safely.

File: src/common/src/array/arrow/arrow_impl.rs (hunk 15 above)
Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:608]: Actually, we have encountered three issues regarding the conversion of the timestamp type:
1. When converting Arrow's timestamp type to our timestamp, we should decide whether to convert it to `timestamp` or `timestampz` based on whether there is a timezone (none or some).
2. When converting Arrow's timestamp type to our timestamp, only microseconds were converted, while the other three time units were not converted.
3. In type comparisons, Arrow's `timestamp(_, none)` should match with rw’s timestamp, while `timestamp(_, some) `should match with rw’s `timestamptz`. However, previously, apart from microseconds, the other units did not match.

#17201 fix 1, and this pr fix 2 and 3.

Comment (hzxa21) [reply src/common/src/array/arrow/arrow_impl.rs:608]: I understand. Correct me if I am wrong, I think 1 is not fixed completely because we don't use the correct timezone for timestamptz based on the arrow data type.

Comment (wcy-fdu) [reply src/common/src/array/arrow/arrow_impl.rs:608]: Double checked, their is no correctness issue:
When the timezone in Arrow is `Some(_)`, the i64 always stores **UTC time**, which is consistent with PG/RW's timestamptz. Therefore, we only need to distinguish between None and Some(_), as the contents of Some do not affect the actual value; they are merely for display purposes.
refer to https://docs.rs/arrow-schema/53.2.0/arrow_schema/enum.DataType.html#timestamps-with-a-non-empty-timezone
